src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -52,58 +52,144 @@ public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, Checksum solutionCh
                 assetMap = await scope.GetAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
             }
 
-            // We can cancel early, but once the pipe operations are scheduled we rely on both operations running to
-            // avoid deadlocks (the exception handler in 'task1' ensures progress is made in 'task2').
             cancellationToken.ThrowIfCancellationRequested();
-            var mustNotCancelToken = CancellationToken.None;
-
-            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC
-            // pipe. Run two tasks - the first synchronously writes to a local pipe and the second asynchronously
-            // transfers the data to the RPC pipe.
-            //
-            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but
-            // might result in more (non-contiguous) memory allocated for the underlying buffers. The amount of memory
-            // is bounded by the total size of the serialized assets.
-            var localPipe = new Pipe(s_pipeOptionsWithUnlimitedWriterBuffer);
-
-            var task1 = Task.Run(() =>
+
+            using var stream = new PipeWriterStream(pipeWriter);
+            await RemoteHostAssetSerialization.WriteDataAsync(
+                stream, singleAsset, assetMap, serializer, scope.ReplicationContext,
+                solutionChecksum, checksums, cancellationToken).ConfigureAwait(false);
+
+            // Ensure any last data written into the stream makes it into the pipe.
+            await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
@AArnott this is hanging.  Note the reader is running and is inside:

`using var stream = await pipeReader.AsPrebufferedStreamAsync(cancellationToken).ConfigureAwait(false);`

not sure what's up here :-/